pkg/check/cache.go

```diff
@@ -0,0 +1,282 @@
+package check
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/vmware/govmomi/find"
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/property"
+	"github.com/vmware/govmomi/view"
+	"github.com/vmware/govmomi/vim25"
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vim25/types"
+	"k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/klog/v2"
+)
+
+// VSphereCache is an interface to cache of frequently accessed vCenter objects,
+// such as datacenter or datastore.
+type VSphereCache interface {
+	// GetDatacenter returns datacenter by name.
+	GetDatacenter(ctx context.Context, dcName string) (*object.Datacenter, error)
+	// GetDataStore returns datastore by name or by path - both can be used.
+	GetDatastore(ctx context.Context, dcName, dsName string) (*object.Datastore, error)
+	// GetDatastoreByURL returns datastore by URL.
+	GetDatastoreByURL(ctx context.Context, dcName, dsURL string) (mo.Datastore, error)
+	// GetDataStoreMo returns datastore's ManagedObject by name or by path - both can be used.
+	GetDatastoreMo(ctx context.Context, dcName, dsName string) (mo.Datastore, error)
+	// GetDatastoreByURL returns datastore by URL.
+	GetDatastoreMoByReference(ctx context.Context, dcName string, ref types.ManagedObjectReference) (mo.Datastore, error)
+	// GetDatastoreByURL returns all StoragePods in a datacenter
+	GetStoragePods(ctx context.Context) ([]mo.StoragePod, error)
+}
+
+// vSphereCache caches frequently accessed vCenter objects for a single check run.
+type vSphereCache struct {
+	vmClient    *vim25.Client
+	datacenters map[string]*cachedDatacenter
+	storagePods []mo.StoragePod
+	mutex       sync.Mutex
+}
+
+type cachedDatacenter struct {
+	dc         *object.Datacenter
+	dcPath     string
+	datastores []*cachedDatastore
+}
+
+type cachedDatastore struct {
+	dsPath string
+	ds     *object.Datastore
+	dsMo   *mo.Datastore
+}
+
+var _ VSphereCache = &vSphereCache{}
+
+func NewCheckCache(vmClient *vim25.Client) VSphereCache {
+	return &vSphereCache{
+		vmClient:    vmClient,
+		datacenters: make(map[string]*cachedDatacenter),
+	}
+}
+
+func (c *vSphereCache) getDatacenterLocked(ctx context.Context, dcName string) (*cachedDatacenter, error) {
+	if cdc, found := c.datacenters[dcName]; found {
```

Comment:
I was thinking very simple fix to prepare for multiple vcenter release. At least dcName here should not be the sole key. May be it should have vcenter name also. 

We can do other stuff later.

Reply:
We don't have vCenter name available in the detector, at least not in the CheckContext:
https://github.com/openshift/vsphere-problem-detector/blob/f7b70bab0066ca209c3993fbc2d1ecb92b4680b1/pkg/check/interface.go#L68-L78

I find it odd to pass "" as vCenter name just in case it might be handy in the future.

Reply:
We have `VsphereConfig` which should have vCenter name?

Reply:
Yeah, but in the end we will need to pass also extra client for each vCenter and who knows how that will look like.

I think that adding multiple vCenter support to the cache will be exactly the same effort now as when we add it later. And later we will know _how_ we to pass vCenter names + clients around the rest of the detector.

Reply:
hmm, okay lets cross this bridge when we have to.